### PR TITLE
[DDC-3346] Failing test for issue (bad findOneBy behaviour with eager fetch)

### PR DESCRIPTION
**Pull request by scaytrase**
Here is the test for [DDC-3346](http://www.doctrine-project.org/jira/browse/DDC-3346) issue

### CODE REVIEW

File: tests/Doctrine/Tests/ORM/Functional/Ticket/DDC3346Test.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                'Doctrine\Tests\Models\DDC3346\DDC3346Article',
+                'Doctrine\Tests\Models\DDC3346\DDC3346Author',
+            )
+        );
+    }
+
+    public function testFindOneByWithEagerFetch()
+    {
+        $user = new DDC3346Author();
+        $user->name = "Buggy Woogy";
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/DDC3346Test.php]: Can you remove these fields?

File: tests/Doctrine/Tests/Models/DDC3346/DDC3346Article.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    public $id;
+    /**
+     * @ManyToOne(targetEntity="DDC3346Author", inversedBy="articles")
+     * @JoinColumn(name="user_id", referencedColumnName="id")
+     */
+    public $user;
+    /**
+     * @Column(type="text")
+     */
+    public $text;
```
Comment (Ocramius) [inline tests/Doctrine/Tests/Models/DDC3346/DDC3346Article.php]: Remove anything that isn't strictly needed

File: tests/Doctrine/Tests/ORM/Functional/Ticket/DDC3346Test.php [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ */
+class DDC3346Test extends \Doctrine\Tests\OrmFunctionalTestCase
+{
+    public function setUp()
+    {
+        parent::setUp();
+
+        $this->setUpEntitySchema(
+            array(
+                'Doctrine\Tests\Models\DDC3346\DDC3346Article',
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/DDC3346Test.php]: Please use `const CLASSNAME = __CLASS__` in the referenced classes definitions: makes referencing classes much simpler and cleaner ;-)

**Comment by Ocramius** [comment]
Wondering about the resolution path for this: my opinion is that we should just disable the eager loading when a `LIMIT` is provided. Thoughts?

**Comment by scaytrase** [comment]
@Ocramius I've updated test. Does it looks right?

Generally, I see three options:
1. Disable eager loading. This leads to unwanted additional queries (one per each entity association)
2. Fetch entity again if found (do something like ```$entity = $this->find($entity->getId())``` internally). This options will make only one additional request constantly.
3. Make more complex query.

**Comment by Ocramius** [comment]
> Disable eager loading.

This is what I'll try doing. The additional queries are a side-effect that cannot be avoided.

 > Fetch entity again if found

Not really possible, as you'll just hit the identity map in this case. I'll look into it though.

 > Make more complex query

This involves going through the paginator API (2 queries), and we really don't want to do that given the horrible amount of paginator-related issues with mssql and other exotic DBMS.

**Comment by scaytrase** [comment]
I took a look through the sources, and the problem really occurs with any limit set. I've updated the test to respect this.

**Comment by scaytrase** [comment]
Also I realized that offset parameter in ```findOneBy```\```findBy``` will also should lead to failure.

I think that disabling eager loading is dirty workaround, because it would have unattended performance impact.
Third way is more clean and correct IMO, but as far as I can understand you, it is rather hard to implement (I almost don't know what is paginator API).

Second way is something intermediate.

**Comment by Ocramius** [comment]
> I think that disabling eager loading is dirty workaround, because it would have unattended performance impact.

Doing the opposite is actually worse, as the state of the loaded entities will be corrupted, and all event listeners will be broken instead. Patching state by refreshing entities is not a good idea.

If the user configures eager loads, it's his own fault there: a second query is an acceptable tradeoff.

**Comment by Ocramius** [comment]
I have a fix for this, but it's not complete yet. Fix will be applied in 2.5: 2.4.x will probably get an exception message instead (for now).

**Comment by Ocramius** [comment]
Closing, moved to #1280

**Comment by Ocramius** [comment]
This bug does not affect 2.4.x, as the eager loading does not work with `TO_MANY` associations

**Pull request closed without merging**